instance/ansible.py

```diff
@@ -67,59 +68,96 @@ def dict_merge(dict1, dict2):
     return dict1
 
 
-@contextmanager
-def string_to_file_path(string):
+def string_to_file_path(string, root_dir=None):
     """
-    Store a string in a temporary file, to pass on to a third-party shell command as a file parameter
-    Returns the file path string
+    Store a string in a temporary file
     """
-    fd, file_path = mkstemp(text=True)
-    fp = os.fdopen(fd, 'w')
-    fp.write(string)
-    fp.close()
-    yield file_path
-    os.remove(file_path)
+    f = NamedTemporaryFile('w', delete=False, dir=root_dir)
+    f.write(string)
+    f.close()
+    return f.name
 
 
 @contextmanager
-def run_playbook(requirements_path, inventory_str, vars_str, playbook_path, playbook_name, username='root'):
+def create_temp_dir():
     """
-    Runs ansible-playbook in a dedicated venv
+    A context manager that creates a temporary directory, returns it. Directory is deleted upon context manager exit.
+    """
+    try:
+        temp_dir = mkdtemp()
+        yield temp_dir
+    finally:
+        # This is really a "best effort" solution, if something will block the deletion (like file being locked somehow)
+        # this should just move on rather than raise.
+        shutil.rmtree(temp_dir, ignore_errors=True)
```

Comment:
In the end I couldn't find a sensible way to eventual handle errors raised by rmtree. For now I don't believe that this error will occur on production, so I decided to just raise the exception, which will log the error, and terminate the server set. 

Ideal solution would be to implement a sort of logging context that would transparently handle passing server context. And just log these errors, but leave instance state unchanged (if it provisioned don't tear it down). 

For now I'd suggest just raising this exception. This also the same way rmtree errors are handled when [deleting github repository.](https://github.com/open-craft/opencraft/blob/d56624843485f588ba76759fb09f8fe50952684d/instance/repo.py#L41)